### ISSUE TITLE
Be able to produce synthetic data for the whole of the star schema

### ISSUE THREAD

**Issue by mhauru** [alan-turing-institute/sqlsynthgen#29]
To do this, run sqlsynthgen on the existing OMOP schema and make sure that it all goes smoothly.

Optionally, as a stretch goal, we could run on the latest InformDB/EMAP, but that would require figuring out how to run InformDB, which sounds not fun.

**Comment by mhauru** [comment]
By the way, once this is done, we should check performance of sqlsynthgen, at least crudely. I feel like there's an issue to link for this, but I can't find it now.

**Comment by mhauru** [comment]
Successfully ran sqlsynthgen on our version of the Star schema, with some but not all foreign key relations in place. Will try on OMOP as well once we find a version with FK constraints.

**Comment by Iain-S** [comment]
When I run this on an OMOP schema, I get sqlacodegen errors because of the cyclical dependency between the `concept` and `domain` tables. I do not know whether the star schema has such dependencies nor how crucial it is that we support them.

1. One solution would be to create the tables, insert data and then enable the constraints but this would interfere with the current workflow.
1. Another would be to disable fk constraints pre- data insertion and reenable them post-. This doesn't interfere with the workflow but I don't know whether SQL alchemy has any support for it or if it would require raw sql.

**Comment by mhauru** [comment]
Sorry, is it `make-tables`, `create-tables`, or `create-data` that errors? If it's `create-data` might it help if we ran all the inserts within a single transaction?

**Comment by Iain-S** [comment]
It's `create-data` and that transaction idea is a very good one.

**Comment by Iain-S** [comment]
After some investigation, I have found that

1. There are some Postgres SQL [workarounds](https://blog.sql-workbench.eu/post/cyclic-foreign-keys/) to circular imports:
  1. using deferred constraints
  1. using a common table expression

A cursory search makes me think that either of these could be used in our case by either

1. adding `conn.execute(text("set constraints all deferred"))` immediately after beginning the transaction
2. using sqlalchemy's [cte support](https://docs.sqlalchemy.org/en/14/core/tutorial.html#common-table-expressions-cte)

There wouldn't be much harm in adding 1. to every transaction but the bigger issue is that we try to select from the second table before we have inserted into it when we loop over the `sorted_generators` list. I think we would need to make our inserts more circular-constraint-aware if we wanted to incorporate this feature.

**Comment by mhauru** [comment]
Thanks, this sounds really useful. I wonder how common circular references are. I thought they would be of mostly academic interest, but then we just ran into one in our second example case, so maybe I'm wrong.

**Comment by mhauru** [comment]
So currently the question of whether we can replicate what Satellite does depends on whether EMAP has circular foreign keys, right? If not, then we can.

**Comment by Iain-S** [comment]
That's right.

Sometime this week, I'll remove the circular key from the OMOP db and check that sqlsynthgen works without it there.

**Comment by Iain-S** [comment]
Note that Satellite updates and deletes rows in addition to inserting them. This isn't something we've put on the roadmap for sqlsynthgen at the moment.

**Comment by Iain-S** [comment]
Using OMOP as our src db, we need to manually remove the cyclical foreign key constraints between 

- vocabulary <-> concept
- domain <-> concept

Though this works around the SQLAlchemy/SQLACodeGen warning, data insertion fails because `domain.domain_id` is a varchar PK without a default/auto-increment. We try to insert NULL into it assuming that a PK will be auto generated but one isn't.

**Comment by mhauru** [comment]
Confirmed that SSG runs fine on the full Star schema extracted from InformDB.
